Assign entries for voting

jayjay19630 opened the pull request:
### Description

This change allows admin to control when they wish to assign / reassign entries for voting. Doing so will delete all existing vote submissions and essentially 'reboot' the assignment so that each entry has a fair chance of being voted on. 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Code quality improvements

### Checklist

<!-- Please delete options that are not relevant. -->

- [x] I have tested this code
- [ ] I have updated the documentation

RichDom2185 commented on src/pages/academy/groundControl/subcomponents/GroundControlConfigureCell.tsx:
```diff
@@ -117,12 +137,43 @@ const ConfigureCell: React.FC<Props> = ({ handleConfigureAssessment, data }) =>
-                  inline
-                  label="Publish Voting (Coming soon!)"
-                />
+                <div className="current-voting-status">
+                  <InfoSign />
+                  <p className="voting-status-text">
+                    Current Voting Status: Entries have {!isVotingPublished && <b>not </b>} been
+                    assigned
+                  </p>
+                </div>
+                {!confirmAssignEntries ? (
+                  <div className="control-button-container">
+                    <ControlButton
+                      icon={IconNames.RESET}
+                      onClick={onAssignClick}
+                      label={`${!isVotingPublished ? 'Assign' : 'Reassign'} entries for voting`}
+                    />
+                  </div>
+                ) : (
+                  <div className="confirm-assign-voting">
+                    <Icon icon="reset" />
+                    <p className="confirm-assign-text">
+                      Are you sure you want to{' '}
+                      <b>{isVotingPublished ? 're-assign' : 'assign'} entries?</b>
+                    </p>
+                    <ButtonGroup>
+                      <Button small intent="success" onClick={handleConfirmAssign}>
+                        Assign
+                      </Button>
+                      <Button small intent="danger" onClick={handleCancelAssign}>
+                        Cancel
+                      </Button>
+                    </ButtonGroup>
+                  </div>
+                )}
+                {isVotingPublished && (
+                  <p className="reassign-voting-warning">
+                    <b>All existing votes</b> will be <b>deleted</b> upon reassigning entries!
+                  </p>
+                )}
```
Thanks for grouping the state logic together using comments. Perhaps it could be better if we can have a separate sub-subcomponent for this toggle instead for better abstraction? I can imagine once we add more configurations into this cell it can become more complex – the `GroundControlConfigureCell` just be focused on the layout, ideally.

jayjay19630 replied:
On that note, do you think it might be a good idea to abstract out the other switches as well (i.e. team assessment, is publish voting, etc.?) I know they're pretty small components but they still might clutter up the code once you add more and more stuff to the config panel

RichDom2185 replied:
I was thinking about it, and I came to the conclusion that we should just abstract as necessary (i.e. only this component, at least for now) – what do you think?

jayjay19630 replied:
Okay then, sounds good. I'll refactor just the assign entries portion inside a new file.

RichDom2185 commented on src/styles/_groundcontrol.scss:
```diff
@@ -85,6 +85,28 @@
     .publish-voting {
       margin-top: 5px;
     }
+
+    .reassign-voting-warning {
+      font-size: 11px;
+      margin-left: 38px;
+    }
+
+    .confirm-assign-voting,
+    .current-voting-status {
+      max-height: 30px;
+      display: flex;
+      align-items: center;
+      gap: 6px;
+      margin-left: 15px;
+
+      .confirm-assign-text {
+        margin-top: 9px;
+      }
+
+      .voting-status-text {
+        margin-top: 10px;
+      }
+    }
```
On the same vein as the above comment, these classes can be separated into a CSS module with the abstracted component name.

RichDom2185 reviewed (commented):
Just a minor comment on the refactor (abstraction). Otherwise, looks good, thanks!

Approved (in-principle), but as usual, we wait for the backend PR to pass review before merging both at the same time.

jayjay19630 commented:
I've added a new folder called configureControls, which includes a new `AssignEntriesButton.tsx`. If any more components need to be added to the panel, they can be organised inside that configureControls folder.

Also, I've added a new scss module just for these configureControls. Let me know if there are any more issues.

RichDom2185 reviewed (approved):
LGTM, thanks a lot!
